vm/src/stdlib/itertools.rs

```diff
@@ -908,48 +908,177 @@ impl PyItertoolsCombinations {
+    r: Cell<usize>,                      // Size of result tuple
+    exhausted: Cell<bool>,               // Set when the iterator is exhausted
+}
+
+impl PyValue for PyItertoolsPermutations {
+    fn class(vm: &VirtualMachine) -> PyClassRef {
+        vm.class("itertools", "permutations")
+    }
+}
+
+#[pyimpl]
+impl PyItertoolsPermutations {
+    #[pyslot(new)]
+    fn tp_new(
+        cls: PyClassRef,
+        iterable: PyObjectRef,
+        r: OptionalArg<PyObjectRef>,
+        vm: &VirtualMachine,
+    ) -> PyResult<PyRef<Self>> {
+        let iter = get_iter(vm, &iterable)?;
+        let pool = get_all(vm, &iter)?;
+
+        let n = pool.len();
+        // If r is not provided, r == n. If provided, r must be a positive integer, or None.
+        // If None, it behaves the same as if it was not provided.
+        let r = match r.into_option() {
+            Some(r) => {
+                if vm.is_none(&r) {
+                    n
+                } else {
+                    let val = r
+                        .payload::<PyInt>()
+                        .ok_or_else(|| vm.new_type_error("Expected int as r".to_string()))?
+                        .as_bigint();
+
+                    if val.is_negative() {
+                        return Err(vm.new_value_error("r must be non-negative".to_string()));
+                    }
+                    val.to_usize().unwrap()
+                }
             }
+            None => n,
+        };
+
+        PyItertoolsPermutations {
+            pool,
+            indices: RefCell::new((0..n).collect()),
+            cycles: RefCell::new((0..r).map(|i| n - i).collect()),
+            result: RefCell::new(None),
+            r: Cell::new(r),
+            exhausted: Cell::new(r > n),
         }
-        // If no suitable index is found, then the indices are all at
-        // their maximum value and we're done.
-        if !sentinel {
+        .into_ref_with_type(vm, cls)
+    }
+
+    #[pymethod(name = "__iter__")]
+    fn iter(zelf: PyRef<Self>, _vm: &VirtualMachine) -> PyRef<Self> {
+        zelf
+    }
+
+    #[pymethod(name = "__next__")]
+    fn next(&self, vm: &VirtualMachine) -> PyResult {
+        // stop signal
+        if self.exhausted.get() {
+            return Err(new_stop_iteration(vm));
+        }
+
+        let n = self.pool.len();
+        let r = self.r.get();
+
+        if n == 0 {
             self.exhausted.set(true);
+            return Ok(vm.ctx.new_tuple(vec![]));
         }
 
-        // Increment the current index which we know is not at its
-        // maximum.  Then move back to the right setting each index
-        // to its lowest possible value (one higher than the index
-        // to its left -- this maintains the sort order invariant).
-        indices[idx] += 1;
-        for j in idx + 1..r {
-            indices[j] = indices[j - 1] + 1;
+        let mut result = self.result.borrow_mut();
+
```

Comment:
To avoid having to do `as_mut().unwrap()` later, you can do `if let Some(ref mut result) = result { ... } else { ... }` (or match if you want to keep the order it's in)

Reply:
Fixed